File: app/src/main/java/com/mask/customcomponents/decoration/DividerItemDecoration.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mask.customcomponents.decoration;

import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.view.View;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.OrientationHelper;
import androidx.recyclerview.widget.RecyclerView;

/**
 * RecyclerView列表间距
 * <p>
 * Created by lishilin on 2016/11/15
 */
public class DividerItemDecoration extends RecyclerView.ItemDecoration {

    private static final Drawable dividerDefault = new ColorDrawable(0x00000000);
    private static final int dividerSizeDefault = 0;

    private Drawable divider;
    private int dividerWidth;
    private int dividerHeight;

    private boolean includeEdge;// 是否包含边缘(垂直布局 - 是否有上下间距，水平布局 - 是否有左右间距)

    public static DividerItemDecoration newInstance() {
        return newInstance(dividerSizeDefault);
    }

    public static DividerItemDecoration newInstance(int size) {
        return newInstance(size, dividerDefault);
    }

    public static DividerItemDecoration newInstance(int size, @ColorInt int color) {
        return newInstance(size, new ColorDrawable(color));
    }

    public static DividerItemDecoration newInstance(int size, Drawable divider) {
        final DividerItemDecoration decoration = new DividerItemDecoration();
        decoration.setDivider(divider);
        decoration.setDividerWidth(size);
        decoration.setDividerHeight(size);
        return decoration;
    }

    @Override
    public void onDraw(@NonNull Canvas c, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        if (divider == null) {
            return;
        }
        final int orientation = getOrientation(parent);

        if (orientation == OrientationHelper.VERTICAL) {
            drawHorizontalDivider(c, parent);
        } else if (orientation == OrientationHelper.HORIZONTAL) {
            drawVerticalDivider(c, parent);
        }
    }

    @Override
    public void getItemOffsets(@NonNull Rect outRect, @NonNull View view, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        final int orientation = getOrientation(parent);

        if (orientation == OrientationHelper.VERTICAL) {
            if (isFirstView(parent, view) && includeEdge) {// 绘制上间距(当 是第一排 且 包含边缘)
                outRect.top = dividerHeight;
            }

            if (!isLastView(parent, view) || includeEdge) {// 绘制下间距(当 不是最后一排 或 包含边缘)
                outRect.bottom = dividerHeight;
            }
        } else if (orientation == OrientationHelper.HORIZONTAL) {
            if (isFirstView(parent, view) && includeEdge) {// 绘制左间距(当 是第一排 且 包含边缘)
                outRect.left = dividerWidth;
            }

            if (!isLastView(parent, view) || includeEdge) {// 绘制右间距(当 不是最后一排 或 包含边缘)
                outRect.right = dividerWidth;
            }
        }
    }

    /**
     * 获取方向
     *
     * @param parent RecyclerView
     * @return 方向
     */
    private int getOrientation(RecyclerView parent) {
        int orientation = -1;
        RecyclerView.LayoutManager layoutManager = parent.getLayoutManager();
        if (layoutManager instanceof LinearLayoutManager) {
            orientation = ((LinearLayoutManager) layoutManager).getOrientation();
        }
        return orientation;
    }

    /**
     * 绘制纵向线条
     *
     * @param c      Canvas
     * @param parent RecyclerView
     */
    private void drawVerticalDivider(Canvas c, RecyclerView parent) {
        final int childCount = parent.getChildCount();
        final int top = 0;
        final int bottom = parent.getHeight();

        for (int childViewIndex = 0; childViewIndex < childCount; childViewIndex++) {
            final View child = parent.getChildAt(childViewIndex);

            if (isFirstView(parent, child) && includeEdge) {// 绘制左间距(当 是第一排 且 包含边缘)
                int left = child.getLeft() - dividerWidth;
                int right = left + dividerWidth;

                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }

            if (!isLastView(parent, child) || includeEdge) {// 绘制右间距(当 不是最后一排 或 包含边缘)
                int left = child.getRight();
                int right = left + dividerWidth;

                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }
        }
    }

    /**
     * 绘制横向线条
     *
     * @param c      Canvas
     * @param parent RecyclerView
     */
    private void drawHorizontalDivider(Canvas c, RecyclerView parent) {
        final int childCount = parent.getChildCount();

        for (int childViewIndex = 0; childViewIndex < childCount; childViewIndex++) {
            final View child = parent.getChildAt(childViewIndex);
            final int left = 0;
            final int right = parent.getWidth();

            if (isFirstView(parent, child) && includeEdge) {// 绘制上间距(当 是第一排 且 包含边缘)
                int top = 0;
                int bottom = top + dividerHeight;

                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }

            if (!isLastView(parent, child) || includeEdge) {// 绘制下间距(当 不是最后一排 或 包含边缘)
                int top = child.getBottom();
                int bottom = top + dividerHeight;

                divider.setBounds(left, top, right, bottom);
                divider.draw(c);
            }
        }
    }

    /**
     * 是否是第一个
     *
     * @param parent RecyclerView
     * @param view   view
     * @return 是否是第一个
     */
    private boolean isFirstView(RecyclerView parent, View view) {
        final int position = parent.getChildAdapterPosition(view);// item position

        return position == 0;
    }

    /**
     * 是否是最后一个
     *
     * @param parent RecyclerView
     * @param view   view
     * @return 是否是最后一个
     */
    private boolean isLastView(RecyclerView parent, View view) {
        RecyclerView.Adapter adapter = parent.getAdapter();
        final int position = parent.getChildAdapterPosition(view);// item position
        int childCount = adapter == null ? 0 : adapter.getItemCount();// item总数

        return position == childCount - 1;
    }

    public boolean isIncludeEdge() {
        return includeEdge;
    }

    public void setIncludeEdge(boolean includeEdge) {
        this.includeEdge = includeEdge;
    }

    /**
     * Sets the drawable to be used as the divider.
     *
     * @param divider divider
     */
    public void setDivider(Drawable divider) {
        if (divider != null) {
            int intrinsicWidth = divider.getIntrinsicWidth();
            int intrinsicHeight = divider.getIntrinsicHeight();
            if (intrinsicWidth > 0) {
                dividerWidth = intrinsicWidth;
            }
            if (intrinsicHeight > 0) {
                dividerHeight = intrinsicHeight;
            }
        } else {
            dividerWidth = 0;
            dividerHeight = 0;
        }
        this.divider = divider;
    }

    /**
     * Gets the drawable currently used as the divider.
     */
    public Drawable getDivider() {
        return divider;
    }

    /**
     * Sets the divider width, in pixels.
     */
    public void setDividerWidth(int dividerWidth) {
        this.dividerWidth = dividerWidth;
    }

    /**
     * Gets the divider width, in pixels.
     */
    public int getDividerWidth() {
        return dividerWidth;
    }

    /**
     * Sets the divider height, in pixels.
     */
    public void setDividerHeight(int dividerHeight) {
        this.dividerHeight = dividerHeight;
    }

    /**
     * Gets the divider height, in pixels.
     */
    public int getDividerHeight() {
        return dividerHeight;
    }

}

File: app/src/main/java/com/mask/customcomponents/decoration/DividerItemDecoration.kt
package com.mask.customcomponents.decoration

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.view.View
import androidx.annotation.ColorInt
import androidx.core.graphics.drawable.toDrawable
import androidx.core.view.forEach
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.OrientationHelper
import androidx.recyclerview.widget.RecyclerView

/**
 * RecyclerView LinearLayoutManager 列表分隔线
 *
 * Created by lishilin on 2016/11/15
 */
class DividerItemDecoration private constructor(
    divider: Drawable?,
    dividerWidth: Int,
    dividerHeight: Int
) : RecyclerView.ItemDecoration() {

    private val mDivider = divider ?: mDividerDefault
    private val mDividerWidth = dividerWidth
    private val mDividerHeight = dividerHeight

    private val mIgnoredPositionList = mutableListOf<Int>() // 忽略分隔线的位置集合

    var isIncludeEdge = false // 是否包含边缘(垂直布局：是否有上下分隔线； 水平布局：是否有左右分隔线；)

    companion object {

        private val mDividerDefault: Drawable
            get() {
                return Color.TRANSPARENT.toDrawable()
            }

        fun newInstance(size: Int): DividerItemDecoration {
            return newInstance(size, mDividerDefault)
        }

        fun newInstance(size: Int, @ColorInt color: Int): DividerItemDecoration {
            return newInstance(size, color.toDrawable())
        }

        fun newInstance(size: Int, divider: Drawable?): DividerItemDecoration {
            return DividerItemDecoration(divider, size, size)
        }
    }

    override fun getItemOffsets(
        outRect: Rect,
        view: View,
        parent: RecyclerView,
        state: RecyclerView.State
    ) {
        when (getOrientation(parent)) {
            OrientationHelper.VERTICAL -> {
                getItemOffsetsByVertical(outRect, view, parent)
            }

            OrientationHelper.HORIZONTAL -> {
                getItemOffsetsByHorizontal(outRect, view, parent)
            }
        }
    }

    override fun onDraw(canvas: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        if (mDivider.alpha == 0) {
            return
        }
        when (getOrientation(parent)) {
            OrientationHelper.VERTICAL -> {
                drawDividerByVertical(canvas, parent)
            }

            OrientationHelper.HORIZONTAL -> {
                drawDividerByHorizontal(canvas, parent)
            }
        }
    }

    /**
     * 获取 偏移量(垂直布局)
     */
    private fun getItemOffsetsByVertical(outRect: Rect, view: View, parent: RecyclerView) {
        // 分隔线偏移量 上
        if (isRequiredDrawDividerLeftOrTop(parent, view)) {
            outRect.top = mDividerHeight
        }

        // 分隔线偏移量 下
        if (isRequiredDrawDividerRightOrBottom(parent, view)) {
            outRect.bottom = mDividerHeight
        }
    }

    /**
     * 获取 偏移量(水平布局)
     */
    private fun getItemOffsetsByHorizontal(outRect: Rect, view: View, parent: RecyclerView) {
        // 分隔线偏移量 左
        if (isRequiredDrawDividerLeftOrTop(parent, view)) {
            outRect.left = mDividerWidth
        }

        // 分隔线偏移量 右
        if (isRequiredDrawDividerRightOrBottom(parent, view)) {
            outRect.right = mDividerWidth
        }
    }

    /**
     * 绘制 分隔线(垂直布局)
     */
    private fun drawDividerByVertical(canvas: Canvas, parent: RecyclerView) {
        val left = 0
        val right = parent.width
        parent.forEach { view ->
            // 绘制上分隔线
            if (isRequiredDrawDividerLeftOrTop(parent, view)) {
                val top = 0
                val bottom = top + mDividerHeight

                mDivider.setBounds(left, top, right, bottom)
                mDivider.draw(canvas)
            }

            // 绘制下分隔线
            if (isRequiredDrawDividerRightOrBottom(parent, view)) {
                val top = view.bottom
                val bottom = top + mDividerHeight

                mDivider.setBounds(left, top, right, bottom)
                mDivider.draw(canvas)
            }
        }
    }

    /**
     * 绘制 分隔线(水平布局)
     */
    private fun drawDividerByHorizontal(canvas: Canvas, parent: RecyclerView) {
        val top = 0
        val bottom = parent.height
        parent.forEach { view ->
            // 绘制左分隔线
            if (isRequiredDrawDividerLeftOrTop(parent, view)) {
                val left = view.left - mDividerWidth
                val right = left + mDividerWidth

                mDivider.setBounds(left, top, right, bottom)
                mDivider.draw(canvas)
            }

            // 绘制右分隔线
            if (isRequiredDrawDividerRightOrBottom(parent, view)) {
                val left = view.right
                val right = left + mDividerWidth

                mDivider.setBounds(left, top, right, bottom)
                mDivider.draw(canvas)
            }
        }
    }

    /**
     * 获取 方向
     */
    private fun getOrientation(parent: RecyclerView): Int {
        return when (val layoutManager = parent.layoutManager) {
            is LinearLayoutManager -> {
                layoutManager.orientation
            }

            else -> {
                -1
            }
        }
    }

    /**
     * 获取 位置
     */
    private fun getViewPosition(parent: RecyclerView, view: View): Int {
        return parent.getChildLayoutPosition(view)
    }

    /**
     * 是否是 第一个
     */
    private fun isViewFirst(parent: RecyclerView, view: View): Boolean {
        val position = getViewPosition(parent, view)
        return position == 0
    }

    /**
     * 是否是 最后一个
     */
    private fun isViewLast(parent: RecyclerView, view: View): Boolean {
        val position = getViewPosition(parent, view)
        val childCount = parent.adapter?.itemCount ?: 0 // item总数
        return position == childCount - 1
    }

    /**
     * 是否是 忽略的位置
     */
    private fun isViewIgnored(parent: RecyclerView, view: View): Boolean {
        val position = getViewPosition(parent, view)
        return mIgnoredPositionList.contains(position)
    }

    /**
     * 是否需要 绘制分隔线(左/上)
     *
     * 是第一排 且 包含边缘
     */
    private fun isRequiredDrawDividerLeftOrTop(parent: RecyclerView, view: View): Boolean {
        val isIncludeEdge = isIncludeEdge
        val isViewFirst = isViewFirst(parent, view)
        return isIncludeEdge && isViewFirst
    }

    /**
     * 是否需要 绘制分隔线(右/下)
     *
     * (不是最后一排 或 包含边缘) 且 没有忽略
     */
    private fun isRequiredDrawDividerRightOrBottom(parent: RecyclerView, view: View): Boolean {
        val isIncludeEdge = isIncludeEdge
        val isViewLast = isViewLast(parent, view)
        val isViewIgnored = isViewIgnored(parent, view)
        return (isIncludeEdge || !isViewLast) && !isViewIgnored
    }

    /************************************************************ S 外部调用 ************************************************************/

    fun clearIgnoredPositionList() {
        mIgnoredPositionList.clear()
    }

    fun setIgnoredPositionList(ignoredPositionList: List<Int>?) {
        clearIgnoredPositionList()
        if (!ignoredPositionList.isNullOrEmpty()) {
            mIgnoredPositionList.addAll(ignoredPositionList)
        }
    }

    /************************************************************ E 外部调用 ************************************************************/
}